hive/agents/td3.py

```diff
@@ -0,0 +1,424 @@
+        self._replay_buffer.add(**self.preprocess_update_info(update_info))
+
+        # Update the agent based on a sample batch from the replay buffer.
+        if (
+            self._learn_schedule.update()
+            and self._replay_buffer.size() > 0
+            and self._update_schedule.update()
+        ):
+            batch = self._replay_buffer.sample(batch_size=self._batch_size)
+            (
+                current_state_inputs,
+                next_state_inputs,
+                batch,
+            ) = self.preprocess_update_batch(batch)
+            with torch.no_grad():
+                noise = torch.randn_like(batch["action"]) * self._target_noise
+                noise = torch.clamp(
+                    noise, -self._target_noise_clip, self._target_noise_clip
+                )
+                next_actions = torch.clamp(
+                    self._target_actor(*next_state_inputs) + noise, -1, 1
+                )
+                next_q_vals = torch.cat(
+                    self._target_critic(*next_state_inputs, next_actions), dim=1
+                )
+                next_q_vals, _ = torch.min(next_q_vals, dim=1, keepdim=True)
+                target_q_values = (
+                    batch["reward"][:, None]
+                    + (1 - batch["done"][:, None]) * self._discount_rate * next_q_vals
+                )
+
+            # Critic losses
+            pred_qvals = self._critic(*current_state_inputs, batch["action"])
+            critic_loss = sum(
+                [mse_loss(qvals, target_q_values) for qvals in pred_qvals]
+            )
+            self._critic_optimizer.zero_grad()
+            critic_loss.backward()
+            if self._grad_clip is not None:
+                torch.nn.utils.clip_grad_norm_(
+                    self._critic.parameters(), self._grad_clip
+                )
+            self._critic_optimizer.step()
+            if self._logger.update_step(self._timescale):
+                self._logger.log_scalar("critic_loss", critic_loss, self._timescale)
+
+            # Update policy with policy delay
+            if self._policy_update_schedule.update():
+                actor_loss = -torch.mean(
+                    self._critic.q1(
+                        *current_state_inputs, self._actor(*current_state_inputs)
+                    )
+                )
+                self._actor_optimizer.zero_grad()
+                actor_loss.backward()
+                if self._grad_clip is not None:
+                    torch.nn.utils.clip_grad_norm_(
+                        self._actor.parameters(), self._grad_clip
+                    )
+                self._actor_optimizer.step()
+                self._update_target()
+                if self._logger.should_log(self._timescale):
+                    self._logger.log_scalar("actor_loss", actor_loss, self._timescale)
+
+    def _update_target(self):
+        """Update the target network."""
+        for (network, target_network) in [
```

Comment:
Should we add flags for hard/soft updates? Some works actually use them in their hyperparameter search